File: 00简单工厂模式/simple-factory-luffycity.py
# -*- coding:utf-8 -*-  

from abc import ABCMeta, abstractclassmethod 

class Payment(metaclass=ABCMeta):
    @abstractclassmethod
    def pay(self, money):
        pass

class Alipay(Payment):
    def __init__(self, use_huabei=False):
        self.use_huabei = use_huabei

    def pay(self, money):
        if self.use_huabei:
            print("花呗支付%d元" % money)
        else:
            print("支付宝余额支付%d元" % money)

class WechatPay(Payment):
    def pay(self, money):
        print("微信支付%d元" % money)




class PaymentFactory:
    def create_payment(self, method):
        if method == 'alipay':
            return Alipay()
        elif method == 'wechat':
            return WechatPay()
        elif method == 'huabei':
            return Alipay(use_huabei=True)
        else:
            return TypeError("No such payment named %s" % method)


# client
pf = PaymentFactory()
# p = pf.create_payment('huabei')
p = pf.create_payment('alipay')
p.pay(100)

File: 00简单工厂模式/simple-factory-luffycity.go
package main

import "fmt"

type Payment interface {
	Pay(money string) string
}

type Alipay struct{}

func (*Alipay) Pay(money string) string {
	return "支付宝余额支付元:" + " " + money
}

type WechatPay struct{}

func (*WechatPay) Pay(money string) string {
	return "微信余额支付元:" + " " + money
}

func PaymentFactory(method string) Payment {
	if method == "alipay" {
		return &Alipay{}
	} else if method == "wechat" {
		return &WechatPay{}
	} else {
		return nil
	}
}

func main() {
	method := "alipay"
	money := "100"
	pay := PaymentFactory(method)
	m := pay.Pay(money)
	fmt.Println(m)
}
